src/libcore/iter.rs

```diff
@@ -115,7 +119,10 @@ pub trait Iterator {
     #[inline]
     #[stable(feature = "rust1", since = "1.0.0")]
     fn count(self) -> usize where Self: Sized {
-        self.fold(0, |cnt, _x| cnt + 1)
+        self.fold(0, |cnt, _| match cnt.checked_add(1) {
+            Some(c) => c,
+            None => panic!("overflow while counting the elements of an iterator"),
```

Comment:
Overflow should be treated as an error and it should be treated as such, we only don't do it in `enumerate` for performance reasons. I see two possibilities for this method, namely returning a `Result` or panicking in the error case. I think that silent errors for consistency with the performance-critical method are not useful.